sbol_utilities/helper_functions.py

```diff
@@ -277,6 +282,78 @@ def has_plasmid_role(x):
-        #return is_plasmid(obj.instance_of.lookup())
     else:
         return False
+
+
+class SBOL3PassiveVisitor:
+    """This base class provides a do-nothing method for every SBOL3 visit type.
+    This allows subclasses to override for only the parts they want to act on"""
+
+    def visit_activity(self, _): pass
+    def visit_agent(self, _): pass
+    def visit_association(self, _): pass
+    def visit_attachment(self, _): pass
+    def visit_binary_prefix(self, _): pass
+    def visit_collection(self, _): pass
+    def visit_combinatorial_derivation(self, _): pass
+    def visit_component(self, _): pass
+    def visit_component_reference(self, _): pass
+    def visit_constraint(self, _): pass
+    def visit_cut(self, _): pass
+    def visit_document(self): pass
+    def visit_entire_sequence(self, _): pass
+    def visit_experiment(self, _): pass
+    def visit_experimental_data(self, _): pass
+    def visit_externally_defined(self, _): pass
+    def visit_implementation(self, _): pass
+    def visit_interaction(self, _): pass
+    def visit_interface(self, _): pass
+    def visit_local_sub_component(self, _): pass
+    def visit_measure(self, _): pass
+    def visit_model(self, _): pass
+    def visit_participation(self, _): pass
+    def visit_plan(self, _): pass
+    def visit_prefixed_unit(self, _): pass
+    def visit_range(self, _): pass
+    def visit_si_prefix(self, _): pass
+    def visit_sequence(self, _): pass
+    def visit_sequence_feature(self, _): pass
+    def visit_singular_unit(self, _): pass
+    def visit_sub_component(self, _): pass
+    def visit_unit_division(self, _): pass
+    def visit_unit_exponentiation(self, _): pass
+    def visit_unit_multiplication(self, _): pass
+    def visit_usage(self, _): pass
+    def visit_variable_feature(self, _): pass
+
+
+def outgoing_links(doc: sbol3.Document) -> set[URIRef]:
+    """Given a document, determine the set of links to objects not in the document
+
+    :param doc: an SBOL document
+    :return: set of URIs for objects not contained in the document
+    """
+    # build a cache and look for all references that cannot be resolved
+    def collector(obj: sbol3.Identified):
+        # Collect all ReferencedURI values in properties:
+        references = []
+        for pv in obj.__dict__.values():
+            if isinstance(pv, ReferencedObjectList):
+                references.extend([v for v in pv if isinstance(v, ReferencedURI)])
+            elif isinstance(pv, ReferencedObjectSingleton):
+                references.append(pv.get())
+
+        # Check whether or not the references resolve
+        for r in references:
+            try:
+                _ = find_top_level(r)
+            except TopLevelNotFound:
+                outgoing.add(str(r))
+            except ValueError:
+                pass  # ignore references to child objects
+
+    outgoing = set()
+    with cached_references(doc):
+        doc.traverse(collector)
+    return outgoing
```

Comment:
Suggestion: add a newline at the end of the file
```suggestion
    return outgoing

```

Reply:
Fixed.